server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,392 @@
+       HOther detail -> do
+         let errMsg = (TBS.fromLBS $ J.encode detail)
+         mkInvo se 500 decodedHeaders errMsg []
+  liftIO $
+    runExceptT $
+    Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) $ do
+      insertInvocation invocation
+      retryOrMarkError se err
+
+retryOrMarkError :: ScheduledEvent 'SE_FULL -> HTTPErr -> Q.TxE QErr ()
+retryOrMarkError se@ScheduledEvent{..} err = do
+  let mretryHeader = getRetryAfterHeaderFromHTTPErr err
+      mretryHeaderSeconds = join $ parseRetryHeaderValue <$> mretryHeader
+      triesExhausted = seTries >= rcstNumRetries seRetryConf
+      noRetryHeader = isNothing mretryHeaderSeconds
+  -- current_try = tries + 1 , allowed_total_tries = rcNumRetries retryConf + 1
+  if triesExhausted && noRetryHeader
+    then do
+      markError
+    else do
+      currentTime <- liftIO getCurrentTime
+      let delay = fromMaybe (rcstIntervalSec seRetryConf) mretryHeaderSeconds
+          diff = fromIntegral delay
+          retryTime = addUTCTime diff currentTime
+      setRetry se retryTime
+  where
+    markError =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET error = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity seId) True
+
+processSuccess :: (MonadIO m) => Q.PGPool -> ScheduledEvent 'SE_FULL -> HTTPResp -> m (Either QErr ())
+processSuccess pgpool se resp = do
+  let respBody = hrsBody resp
+      respHeaders = hrsHeaders resp
+      respStatus = hrsStatus resp
+      decodedHeaders = []
+      invocation = mkInvo se respStatus decodedHeaders respBody respHeaders
+  liftIO $
+    runExceptT $
+    Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) $ do
+      insertInvocation invocation
+      markSuccess
+  where
+    markSuccess =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET delivered = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity $ seId se) True
+
+processDead :: (MonadIO m) => Q.PGPool -> ScheduledEvent 'SE_FULL -> m (Either QErr ())
+processDead pgpool se =
+  liftIO $
+  runExceptT $ Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) markDead
+  where
+    markDead =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+          UPDATE hdb_catalog.hdb_scheduled_events
+          SET dead = 't', locked = 'f'
+          WHERE id = $1
+        |] (Identity $ seId se) False
+
+setRetry :: ScheduledEvent 'SE_FULL -> UTCTime -> Q.TxE QErr ()
+setRetry se time =
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+          UPDATE hdb_catalog.hdb_scheduled_events
+          SET next_retry_at = $1, locked = 'f'
+          WHERE id = $2
+          |] (time, seId se) True
+
+mkInvo
+  :: ScheduledEvent 'SE_FULL -> Int -> [HeaderConf] -> TBS.TByteString -> [HeaderConf]
+  -> Invocation
+mkInvo se status reqHeaders respBody respHeaders
```

Comment:
I think this should probably be renamed to `mkInvocation`; the abbreviation seems extreme.